### ISSUE TITLE
Reconsider how we handle required form fields

### ISSUE THREAD

**Issue by jamesfan** [Workday/canvas-kit#845]
_While adding support for `aria-required` to the Labs Select (#797), a number of other questions came up regarding how we handle required form fields in Canvas Kit in general. This issue summarizes those questions._

## 🔍 Spike / Exploration

Currently, FormField accepts a `required` prop. If `required` is set to `true`, FormField renders `<abbr title="required">*</abbr>` inside of its `<label>`. This is _not_ a recommended approach for signaling to assistive technology that this field is required. From @mathalete:

> The `<abbr>` tag is used to expand the definition of an abbreviated word, this approach isn't recommended. For example - VoiceOver simply does not announce the title attribute content on a abbr
>
> if the ask here is to robustly include the word 'required' in the `<label>` then adding visually hidden text would be the way to do so

Peter suggested the following markup (where `.sr` is the accessible hiding class):
```html
<label for="foo">Name <span aria-hidden="true">*</span> <span class="sr">Required</span></label>
```

On a related note, we also need to determine if we need to add special handling for `aria-required` to FormField. MDN states `aria-required` as a [useful fallback](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-required_attribute) for user agents that do not yet support HTML5, although a cursory Google search suggests that `required` may be sufficient ([source 1](https://developer.paciellogroup.com/blog/2019/02/required-attribute-requirements/), [source 2](https://stackoverflow.com/a/37975985)).

## What questions would be answered once this spike was completed?

1. What was the original reasoning behind adding `<abbr title="required">` for required FormFields? If its sole purpose is to signal to assistive technology that the FormField is required, should we instead implement the suggested fix above where "required" is visually hidden? If not, what would be the recommended solution?
2. Does FormField need to handle `aria-required` as well? If so, do we need to handle it any differently than `required`? 
3. ~Regardless of whether or not we add support for `aria-required`, if we implement the suggested fix referenced in (1) (or whatever solution we arrive at), do we need to retain the code we added to the Labs Select in #797 to special-case handling of `aria-required` or is it no longer needed?~ _Edit: The Labs Select is a special case among input components. Investigate this separately._
4. Do we need any additional handling of `required` and/or `aria-required` in the input components themselves?

**Comment by NicholasBoll** [comment]
Is there any serious objections to the following?

```html
<label for="foo">Name <span aria-hidden="true">*</span></label>
<input type="text" id="foo" name="foo" required />
```

The problem with the `<span className="sr-only">Required</span>` is it will have to be separately translated. Some languages would prepend the "required" translation vs the end which means the application developers will have to be aware of this nuance and we'll have to allow for a much more complicated translation API. Also, wouldn't this mean that "Required" could be announced to a user twice? Once for the `required` on the input and once for the label?

If our supported SR matrix can handle the above markup, that's what we should do. This is one of the recommended Deque options: https://dequeuniversity.com/tips/identify-required-elements

**Comment by mathalete** [comment]
Nope https://workday-dev.slack.com/archives/G017XAPL2KB/p1595586662050700?thread_ts=1595458627.033000&cid=G017XAPL2KB

**Comment by aaroncalderon** [comment]
> Nope https://workday-dev.slack.com/archives/G017XAPL2KB/p1595586662050700?thread_ts=1595458627.033000&cid=G017XAPL2KB there's a few ways to this for sure.

Non Workday Devs cannot see the linked content.

**Comment by NicholasBoll** [comment]
> No objections.

Just to confirm, I built a code sandbox to test across a range of screen readers: https://yuz66.csb.app/

Source code can be found here: https://codesandbox.io/s/yuz66

**Comment by jamesfan** [comment]
Do we also need to apply `aria-required` to the `<input>` in Nicholas' proposed solution? So, something like this:
```
<label for="foo">Name <span aria-hidden="true">*</span></label>
<input type="text" id="foo" name="foo" required aria-required="true" />
```

**Comment by jamesfan** [comment]
Closing this spike.

Given the following JSX...
```
<FormField inputId="foo" required={true} label="Name">
  <TextInput />
</FormField>
```
...the current implementation of FormField renders the following markup:
```
<div>
  <label for="foo" required>
    Name
    <abbr title="required">*</abbr>
  </label>
  <div>
    <input type="text" id="foo" />
  </div>
</div>
```

Given the findings of this spike, the conclusion is to modify FormField to render the following markup proposed by @NicholasBoll and approved by @mathalete:
```
<div>
  <label for="foo">
    Name
    <span aria-hidden="true">*</span>
  </label>
  <div>
    <input type="text" id="foo" required />
  </div>
</div>
```

The `required` attribute will need to be applied to the form input rather than the `label`, and `<abbr title="required">` will need to be replaced with `<span aria-hidden="true">`.

Note that special handling may be required for some input components when the `required` prop is forwarded to the input. For example, consider the Canvas Kit Labs Select which is currently rendered as a `button`: given that `required` is _not_ a valid attribute for a `button`, we need to add special handling for the `required` prop rather than simply applying the `required` attribute to the underlying HTML element like we can for most input components.
